src/path.cr

```diff
@@ -0,0 +1,849 @@
+  end
+
+  # Appends the given *parts* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"].join("bar", "baz")       # => Path["foo/bar/baz"]
+  # Path["foo/"].join("/bar/", "/baz")   # => Path["foo/bar/baz"]
+  # Path["/foo/"].join("/bar/", "/baz/") # => Path["/foo/bar/baz/"]
+  # ```
+  def join(*parts) : Path
+    join parts
+  end
+
+  # Appends the given *parts* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"].join("bar", "baz")           # => Path["foo/bar/baz"]
+  # Path["foo/"].join(Path["/bar/", "/baz"]) # => Path["foo/bar/baz"]
+  # Path["/foo/"].join("/bar/", "/baz/")     # => Path["/foo/bar/baz/"]
+  # ```
+  def join(parts : Enumerable) : Path
+    new_name = String.build do |str|
+      str << @name
+      last_ended_with_separator = @type.ends_with_separator?(@name)
+
+      parts.each_with_index do |part, index|
+        case part
+        when Path
+          part = part.@name
+        else
+          part = part.to_s
+          part.check_no_null_byte
+        end
+
+        if part.empty?
+          if index == parts.size - 1
+            str << @type.separator unless last_ended_with_separator
+            last_ended_with_separator = true
+          else
+            last_ended_with_separator = false
+          end
+
+          next
+        end
+
+        byte_start = 0
+        byte_count = part.bytesize
+
+        case {@type.starts_with_separator?(part), last_ended_with_separator}
+        when {true, true}
+          byte_start += 1
+          byte_count -= 1
+        when {false, false}
+          str << @type.separator unless str.bytesize == 0
+        end
+
+        last_ended_with_separator = @type.ends_with_separator?(part)
+
+        str.write part.unsafe_byte_slice(byte_start, byte_count)
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Appends the given *part* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"] / "bar" / "baz"     # => Path["foo/bar/baz"]
+  # Path["foo/"] / Path["/bar/baz"] # => Path["foo/bar/baz"]
+  # ```
+  def /(part : Path | String) : Path
+    join(part)
+  end
+
+  # :nodoc:
+  def <<(part : Path | String) : Path
+    @name = join(part).@name
+    self
+  end
+
+  # Resolves path *name* in this path's parent directory.
+  #
+  # Raises `Path::Error` if `#parent` is `nil`.
+  def sibling(name : Path | String) : Path?
+    if parent = self.parent
+      parent.join(name)
+    else
+      raise Error.new("Can't resolve sibling for a path without parent directory")
+    end
+  end
+
+  # Compares this path to *other*.
+  # The comparison is performed strictly lexically: `foo` and `./foo` are *not*
+  # treated as equal. To compare paths semantically, they need to be normalized
+  # and converted to the same flavour.
+  #
+  # Comparison is case-sensitive for POSIX paths and case-insensitive for
+  # Windows paths.
+  def <=>(other : Path)
+    ord = @name.compare(other.@name, case_insensitive: windows?)
```

Comment:
Since case sensitivity depends on partition rather than platform I am not sure the default should change depending on the platform. I think is safer to be case insensitive by default.

Did you base this behavior in other implementations?

Reply:
This is based on Python's pathlib (https://docs.python.org/3/library/pathlib.html#general-properties). It's a difficult field, and I'm not entirely sure about this either.

You're right that case sensitivity depends on the file system. Even NTFS supports that, but Windows itself can't really use it. I think it's fairly reasonable to treat paths as case-insensitive on Windows because of that.

I'd consider case-sensitive the safe default. That's what you get when there is no `Path` type and you use string comparison.

Reply:
It's safe to assume - for now at least - that on windows everything is case insensitive and the opposite on linux. It's the closest to the truth you're going to get.

Reply:
Ok. Can be resolved.

Reply:
Ok. Marking as Resolved